File: Java/src/ProbabilityofaTwoBoxesHavingTheSameNumberofDistinctBalls.java
import java.util.ArrayList;

public class ProbabilityofaTwoBoxesHavingTheSameNumberofDistinctBalls {
	
	private int[] fac = {1, 1, 2, 6, 24, 120, 720};
	private double aux = 0;
	public double getProbability(int[] balls) {
		int sum = 0;
		for(int i : balls) sum += i;
		double total = 1;
		for(int i = 1; i <= sum; i++) total *= i;
		for(int i : balls) total /= fac[i];
		
		int tar = sum / 2;
		aux = fac(tar);
		aux = aux * aux;
		int[] a = new int[balls.length];
		int[] b = new int[balls.length];
		
		double valid = dfs(0, balls, a, b, 0, 0, tar, 0, 0);
		return valid / total;
    }
	
	
	private double dfs(int i, int[] balls, int[] a, int[] b, int sa, int sb, int tar, int ca, int cb) {
		double res = 0;
		
		for(int n = 0; n <= balls[i]; n++) {
			a[i] = n;
			b[i] = balls[i]-n;
			int nca = ca;
			if(a[i] > 0) nca++;
			int ncb = cb;
			if(b[i] > 0) ncb++;
			
			int nsa = sa + a[i];
			int nsb = sb + b[i];
			
			if(nsa > tar || nsb > tar) continue;
			
			if(i == balls.length-1) {
				if(nsa != nsb || nca != ncb) continue;
				double cur = aux;
				for(int c : a) 
					if(c > 1) cur = cur / fac[c];
				for(int c : b)
					if(c > 1) cur = cur / fac[c];
				
				return cur;
			}else {
				res += dfs(i+1, balls, a, b, nsa, nsb, tar, nca, ncb);
			}
			
		}
		return res;
	}
	
	private double fac(int n) {
		double res = 1;
		for(int i = 1; i <= n; i++) res *= i;
		System.out.println(res);
		return res;
	}
	
	public static void main(String[] args) {
		ProbabilityofaTwoBoxesHavingTheSameNumberofDistinctBalls test = new ProbabilityofaTwoBoxesHavingTheSameNumberofDistinctBalls();
		System.out.println(test.getProbability(new int [] {6,6,6,6,6,6}));
	}
	
	
	
	
}

File: Python/ProbabilityofaTwoBoxesHavingTheSameNumberofDistinctBalls.py
import math
from typing import List

class Solution:
    def getProbability(self, balls: List[int]) -> float:
        k = len(balls)
        sum = 0
        for b in balls:
            sum += b
        total = math.factorial(sum)
        self.tar = sum / 2
        for b in balls:
            total = total / math.factorial(b)
        a = [0]*k;
        b = [0]*k;
        self.aux = math.factorial(self.tar)
        self.aux = self.aux * self.aux
        valid = self.dfs(0, balls, a, b, 0, 0, 0, 0)
        return valid / total

    def dfs(self, i, balls, a, b, sa, sb, ca, cb):
        res = 0
        for n in range(balls[i]+1):
            a[i] = n
            b[i] = balls[i]-n
            nca = ca
            if n > 0:
                nca += 1
            ncb = cb
            if balls[i]-n > 0:
                ncb += 1

            nsa = sa+n
            nsb = sb+balls[i]-n

            if nsa > self.tar or nsb > self.tar:
                continue

            if abs(nca-ncb) > len(balls)-1-i:
                continue

            if i == len(balls)-1:
                if nsa != nsb or nca != ncb:
                    continue
                cur = self.aux

                for c in a:
                    if c > 1:
                        cur = cur / math.factorial(c)

                for c in b:
                    if c > 1:
                        cur = cur / math.factorial(c)

                return cur

            else:
                res += self.dfs(i+1, balls, a, b, nsa, nsb, nca, ncb)
        return res

test = Solution()
print(test.getProbability([6,6,6,6,6,6]))




